### ISSUE TITLE
race-condition in optimizer.py

### ISSUE THREAD

**Issue by szutenberg** [google-research/bert#1050]
Hi,

Please have a look at optimizer.py:
```
76  train_op = optimizer.apply_gradients(
77      zip(grads, tvars), global_step=global_step)
78
79  # Normally the global step update is done inside of `apply_gradients`.
80  # However, `AdamWeightDecayOptimizer` doesn't do this. But if you use
81  # a different optimizer, you should probably take this line out.
82  new_global_step = global_step + 1
83  train_op = tf.group(train_op, [global_step.assign(new_global_step)])
84  return train_op
```

global_step is in the same time incremented and passed to the optimizer. That causes race condition and warnings from TensorFlow:
`WARNING:tensorflow:It seems that global step (tf.train.get_global_step) has not been increased. Current value (could be stable): 123 vs previous value: 123. You could increase the global step by passing tf.train.get_global_step() to Optimizer.apply_gradients or Optimizer.minimize.`

It is confusing for many users:
https://github.com/NVIDIA/DeepLearningExamples/issues/80
https://github.com/google-research/bert/issues/915

There are some suggestions related to NaN loss. I've analyzed computation graphs and there is no such mechanism implemented. It seems that all problems with global_step are caused by the race condition explained above. 

Possible solution: use tf.control_dependencies instead of tf.group.

Best,
Michał

**Comment by agarwalishan** [comment]
> Hi,
> 
> Please have a look at optimizer.py:
> 
> ```
> 76  train_op = optimizer.apply_gradients(
> 77      zip(grads, tvars), global_step=global_step)
> 78
> 79  # Normally the global step update is done inside of `apply_gradients`.
> 80  # However, `AdamWeightDecayOptimizer` doesn't do this. But if you use
> 81  # a different optimizer, you should probably take this line out.
> 82  new_global_step = global_step + 1
> 83  train_op = tf.group(train_op, [global_step.assign(new_global_step)])
> 84  return train_op
> ```
> 
> global_step is in the same time incremented and passed to the optimizer. That causes race condition and warnings from TensorFlow:
> `WARNING:tensorflow:It seems that global step (tf.train.get_global_step) has not been increased. Current value (could be stable): 123 vs previous value: 123. You could increase the global step by passing tf.train.get_global_step() to Optimizer.apply_gradients or Optimizer.minimize.`
> 
> It is confusing for many users:
> [NVIDIA/DeepLearningExamples#80](https://github.com/NVIDIA/DeepLearningExamples/issues/80)
> #915
> 
> There are some suggestions related to NaN loss. I've analyzed computation graphs and there is no such mechanism implemented. It seems that all problems with global_step are caused by the race condition explained above.
> 
> Possible solution: use tf.control_dependencies instead of tf.group.
> 
> Best,
> Michał

I don't have knowledge of tensorflow and I am facing this issue and I don't know how to modify the above code
Can you please write modified code for this specially what arguments will be passed to tf.control_dependencies?
Thanks in advance...
